Automattic/woocommerce-payments
# Display specific error message instead of a generic one in the checkout block when payments fail

Fixes #3099 

#### Changes proposed in this Pull Request

Throw Exceptions with the error message instead of adding error notices when the payment processing fails.

* Remove `wc_add_notice()` for non-UPE payment methods, and for UPE when the limit is reached
* Throw an Exception with the message previously displayed via `wc_add_notice()`

This makes the error message be displayed both in the regular and in the blocks checkout.

This is the generic error that was displayed before
![image](https://user-images.githubusercontent.com/41606954/139495551-e5603004-258e-45b1-9f79-d339eb11302c.png)

#### Testing instructions

For non-UPE errors:
* Make sure you're using the Blocks checkout
* Add 1 or more products to your cart
* Go to the blocks checkout page
* Pay with a card that will be declined, e.g. 4000 0000 0000 0002 or 4000 0000 0000 0127
* At the top of the page, an error message describing the error should be displayed, instead of the generic "Something went wrong. Please contact us to get assistance." one mentioned in the issue.

Error with the card 4000 0000 0000 0127
![image](https://user-images.githubusercontent.com/41606954/139495350-8dab0907-3d5f-44c0-ba31-da788f5632cd.png)

Error with the card 4000 0000 0000 0002
![image](https://user-images.githubusercontent.com/41606954/139495472-b4cb41fc-e0a8-485a-9d18-9dc7f5cb0236.png)


For UPE's cards:
* Make sure you're using the Blocks checkout
* Add 1 or more products to your cart
* Go to the blocks checkout page
* Attempt 6 times to pay with a card that will be declined, e.g. 4000 0000 0000 0002. This will reach the limiter's limit
* Instead of the generic error message, you should get "Your payment was not processed"
![image](https://user-images.githubusercontent.com/41606954/139496527-20e2a0a4-6021-4fb9-b738-a4d9fefcba8e.png)


-------------------


**Post merge**
